gflownet/envs/crystals.py

```diff
@@ -369,3 +385,16 @@ def _can_produce_neutral_charge(self, state: Optional[List[int]] = None) -> bool
         ]
 
         return any(poss_charge_sum)
+
+    def _can_eos(self, state: List, state_elem: List, n_state_atoms: int):
+        """
+        Returns True if the end of sequence action (EOS) can be chosen from the
+        state passed as an argument. False otherwise.
+        """
+        if n_state_atoms < self.min_atoms:
+            return False
+        if len(state_elem) < self.min_diff_elem:
+            return False
+        if any(r not in state_elem for r in self.required_elements):
```

Comment:
Probably minor because `state_elem` is small but I'm always uncomfortable with `if el in list` vs `if el in set`. Feel free to just resolve without taking any action. I think the overhead of `set` creation is not worth it here.

```python
In [8]: def f(l1, l2):
   ...:     return any(r not in l1 for r in l2)
   ...: 

In [9]: def h(l1, l2):
   ...:     s1 = set(l1)
   ...:     return any(r not in s1 for r in l2)
   ...: 

In [10]: import numpy as np

In [11]: l1, l2 = np.random.randint(0, 100, (2, 30))

In [12]: %timeit f(l1, l2)
4.12 µs ± 90.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [13]: %timeit h(l1, l2)
3.04 µs ± 173 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [14]: l1, l2 = np.random.randint(0, 100, (2, 3000))

In [15]: %timeit f(l1, l2)
14.2 ms ± 3.48 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [16]: %timeit h(l1, l2)
521 µs ± 21.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Reply:
This is part of another PR merged here, but what is the problem with `if el in list`?

Reply:
`if el in list` is `O(len(list))` but `if el in set` is `O(1)` so if you want to check for presence in a list >1 times, it's worth turning it into a `set` first. Obviously the larger the list, and the more you test `in` the more it saves you time

Reply:
https://wiki.python.org/moin/TimeComplexity

Reply:
Oh I see! I knew about this but I haven't incorporated it into my practices. There may be a few places in the repo where we check long lists, so something to review at some point. Thanks! And sorry I didn't understand initially, got confused by something.